evm/src/main/java/org/hyperledger/besu/evm/internal/Words.java

```diff
@@ -91,17 +91,47 @@ public static int numWords(final Bytes input) {
    */
   public static long clampedToLong(final Bytes uint) {
     if (uint.size() <= 8) {
-      long result = uint.toLong();
+      final long result = uint.toLong();
       return result < 0 ? Long.MAX_VALUE : result;
     }
 
-    Bytes trimmed = uint.trimLeadingZeros();
+    final Bytes trimmed = uint.trimLeadingZeros();
     if (trimmed.size() <= 8) {
-      long result = trimmed.toLong();
+      final long result = trimmed.toLong();
       return result < 0 ? Long.MAX_VALUE : result;
     } else {
       // clamp to the largest int.
       return Long.MAX_VALUE;
     }
   }
+
+  /**
+   * Adds a and b, but if an underflow/overflow occurs return the Long max/min value
+   *
+   * @param a first value
+   * @param b second value
+   * @return value of a plus b if no over/underflows or Long.MAX_VALUE/Long.MIN_VALUE otherwise
+   */
+  public static long clampedAdd(final long a, final long b) {
+    try {
+      return Math.addExact(a, b);
+    } catch (final ArithmeticException ae) {
+      return a > 0 ? Long.MAX_VALUE : Long.MIN_VALUE;
```

Comment:
Glad we are clamping these, very defensive. Is it worth it to log something here? When used for gas related calculations, and with blocks aiming for 30M gas in 'em, if these were to actually get clamped it would indicate some pretty big problem that would break elsewhere. Basically when clamping happens, our assumption about longs being big enough for gas becomes invalidated.

Reply:
(a) 30MM gas is ~2^35, long is ~2^63. Any clamping would almost always indicate an out of gas error
(b) [EIP-4803](https://eips.ethereum.org/EIPS/eip-4803) specifies int64 as the constraints for gas, after paying for intrinsic gas the MAX_LONG is an unachievable value, so it's a good proxy for "not enough gas in the world"
(c) The old Gas object was clamping too, just keeping it out of sight.